libs/cardano-ledger-conformance/src/Test/Cardano/Ledger/Conformance.hs

```diff
@@ -1,10 +1,368 @@
-module Test.Cardano.Ledger.Conformance (
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE TypeFamilyDependencies #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
 
-) where
+module Test.Cardano.Ledger.Conformance
+  ( SpecCastable (..)
+  ) where
+import Data.Text (Text)
+import Cardano.Ledger.Shelley.LedgerState (UTxOState (..))
+import qualified Lib as Agda
+import Cardano.Ledger.UTxO (UTxO (..))
+import qualified Data.Map.Strict as Map
+import Cardano.Ledger.TxIn (TxIn (..), TxId (..))
+import Cardano.Ledger.Conway.TxOut (BabbageTxOut (..))
+import Cardano.Ledger.Core (EraTxOut(..)
+  , EraIndependentData, PParams, EraTxBody (..), EraTxAuxData (..), EraTx (..), EraIndependentScriptIntegrity, EraIndependentTxBody, Era)
+import Cardano.Ledger.Coin (Coin (..))
+import Cardano.Ledger.Address (Addr (..), serialiseAddr)
+import Cardano.Ledger.Val (Val(..))
+import Cardano.Ledger.Alonzo.Scripts.Data (Datum(..))
+import Cardano.Ledger.SafeHash (SafeHash, extractHash)
+import Data.Bitraversable (bimapM)
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import Cardano.Crypto.Hash (hashToBytes, Hash)
+import Cardano.Ledger.BaseTypes (
+  SlotNo (..), EpochNo (..), ProtVer (..), getVersion, StrictMaybe, strictMaybeToMaybe, TxIx (..))
+import Cardano.Ledger.Shelley.Rules (UtxoEnv (..), Identity)
+import Cardano.Ledger.Conway.PParams (ConwayPParams (..), DRepVotingThresholds
+  , PoolVotingThresholds)
+import Cardano.Ledger.Alonzo.Scripts (CostModels, Prices, ExUnits (..))
+import Cardano.Ledger.Alonzo.PParams (OrdExUnits (OrdExUnits))
+import Cardano.Ledger.Conway.Core (CoinPerByte (..), EraTxWits (..), vldtTxBodyL, AlonzoEraTxBody (..), EraScript (..))
+import Cardano.Ledger.Alonzo.Tx (AlonzoTx (..))
+import Cardano.Ledger.Alonzo.TxWits (AlonzoTxWits (..), TxDats, Redeemers)
+import Cardano.Ledger.Alonzo (AlonzoTxAuxData, AlonzoScript)
+import Data.Set (Set)
+import qualified Data.Set as Set
+import Data.Sequence.Strict (StrictSeq)
+import Data.Foldable (Foldable(..))
+import Cardano.Ledger.Binary (Sized (..))
+import Cardano.Ledger.Allegra.Scripts (ValidityInterval (..))
+import Cardano.Ledger.Keys (KeyHash (..), VKey (..))
+import Lens.Micro ((^.))
+import Cardano.Ledger.Block (txid)
+import Cardano.Ledger.Keys.WitVKey (WitVKey (..))
+import Cardano.Ledger.Crypto (Crypto (..))
+import Data.Data (Typeable)
+import Cardano.Crypto.DSIGN (DSIGNAlgorithm(..), SignedDSIGN (..))
 
-import Lib (TxBody)
+type SpecTranslationError = Text
 
--- Placeholder so that the compiler wouldn't complain about
--- `cardano-ledger-executable-spec` being a redundant dependency
-_placeholder :: TxBody
-_placeholder = undefined
+class SpecCastable a where
```

Comment:
We'll also need something similar for converting some of the types back from spec rep.

So, maybe name it `ToSpecRep` and another one will be `FromSpecRep`.

`SpecRep` can be a standalone type family, it doesn't have to be part of the type class

Reply:
Yeah, that's great. As long as we are not using Show instance for equality :smile:

Reply:
I was thinking more about this while reviewing the PR and I think it will be problematic if we do comparison on the Agda side. This about test failures that we will get, for example instead of hash values that we generated we will get bunch of integers. So it might be hard to map to the original values that we will need to use for debugging failures.

We don't have solve it in this PR, but some food for though. If we are to do the translation from executable spec back to Ledger code we can provide the current Ledger value and just overwrite only the fields that are present in executable spec. Eg.
```haskell
fromSpecRep :: a -> SpecRep a -> a
```
I don't think it would even have to have a failure mode (i.e. `Either`), since I am sure we can represent all of the executable spec types

Reply:
I thought of this as well and I think I have a better idea, we can talk about this later.